### PR TITLE
Require `belongs_to` by default.

### LINKED ISSUE #18233: belongs_to should default to required: true
Almost every belongs_to declaration seems to be a required association. It's rare that you allow your foreign key to be nil. So let's have required: true be the default and required: false be the optional turn-off.

### PR DESCRIPTION
**Pull request by simi**
Deprecate `required` option in favor of `optional` for `belongs_to`.

closes #18233

### CODE REVIEW

**Comment by simi** [comment]
I wonder what to do with failing test.

```
rails generate scaffold LineItems product:references cart:belongs_to
bundle exec rake db:migrate test`
```

This code fails now, since belongs_to relation is required by default. Should I change generated test for belongs_to to make it pass or should I remove this test since it is not valid with this change?

File: activerecord/lib/active_record/associations/builder/belongs_to.rb [hunk 1]
```diff
@@ -110,5 +110,24 @@ def self.add_destroy_callbacks(model, reflection)
       name = reflection.name
       model.after_destroy lambda { |o| o.association(name).handle_dependency }
     end
+
+    def self.define_validations(model, reflection)
+      if reflection.options.key?(:required)
+        ActiveSupport::Deprecation.warn('`required` option for belongs_to relation is deprecated and replaced with `optional` option.')
```
Comment (simi) [inline activerecord/lib/active_record/associations/builder/belongs_to.rb]: Better info will be needed here. Any ideas?

Comment (dhh) [inline activerecord/lib/active_record/associations/builder/belongs_to.rb]: '`required: true` is now the default for belongs_to and can be removed.' – don't even need to talk about optional here. What they wanted to do is now being done for them.

**Comment by dhh** [comment]
You can kill that test and replace it with a test that tests optional: true.

File: activerecord/lib/active_record/associations/builder/belongs_to.rb (hunk 1 above)
Comment (simi) [inline activerecord/lib/active_record/associations/builder/belongs_to.rb]: What about when `required: false` is specified explicitly? Maybe throw deprecation warning only for `false` value?

Comment (dhh) [inline activerecord/lib/active_record/associations/builder/belongs_to.rb]: That was the old default, which means no client code will have that code, since it doesn’t make any sense.

Comment (simi) [inline activerecord/lib/active_record/associations/builder/belongs_to.rb]: OK, makes sense.

**Comment by simi** [comment]
Should I add also this to generator? `rails g scaffold post author:belongs_to:optional` ?

**Comment by dhh** [comment]
Nah, don’t think that’s necessary. Will be a very uncommon option to use, I think.

**Comment by simi** [comment]
OK I think my work is almost done here. I'll add also CHANGELOG note soon.

**Comment by dhh** [comment]
Awesome. Ping when ready to merge.

**Comment by simi** [comment]
I added CHANGELOG note, but it is a little fuzzy to me. Feel free to ping me with better one.

File: activerecord/CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,9 @@
+*   Require `belongs_to` association by default.
+    This can by configure by `config.active_record.require_belongs_to` option
+    or by `optional: true` option on relation.
+
+    *Josef Šimánek*
```
Comment (dhh) [inline activerecord/CHANGELOG.md]: Let's try with:

`belongs_to` will now trigger a validation error by default if the association is not present. You can turn this off on a per-association basis with `optional: true`. (Note this new default only applies to new Rails apps that will be generated with `config.active_record.require_belongs_to = true`)

**Comment by dhh** [comment]
Let's update the config option to be a bit clearer: `belongs_to_required_by_default = true`

File: activerecord/CHANGELOG.md (hunk 2 above)
Comment (simi) [inline activerecord/CHANGELOG.md]: Actually the final note is not true right now. `config.active_record.require_belongs_to = true` is set directly in railtie, not in application config. Should I move it to new generated `application.rb` and let it `nil` for existing applications? In that case your note will be true.

**Comment by dhh** [comment]
This is also missing the config/initializers/active_record_belongs_to_required_by_default.rb file that's going to set the `belongs_to_required_by_default = true` setting by default for new apps.

**Comment by simi** [comment]
@dhh I mentioned that https://github.com/rails/rails/pull/18937/files#r24953865 already.

**Comment by dhh** [comment]
Ah, we crossed streams. Didn't note that one. Yes, please do go ahead. Thanks!

File: activerecord/lib/active_record/associations/builder/belongs_to.rb [hunk 3]
```diff
@@ -110,5 +110,23 @@ def self.add_destroy_callbacks(model, reflection)
...
+    def self.define_validations(model, reflection)
+      if reflection.options.key?(:required)
+        reflection.options[:optional] = !reflection.options.delete(:required)
+      end
+
+      if reflection.options[:optional].nil?
+        required = model.require_belongs_to
+      else
+        required = !reflection.options[:optional]
+      end
```
Comment (jeremy) [inline activerecord/lib/active_record/associations/builder/belongs_to.rb]: I'm finding the control flow here a little difficult to follow. Consider condensing the conditionals and using the `:optional` option to determine whether to declare the presence validation:
```ruby
# Disallow conflicting options
if reflection.options.key?(:required) && reflection.options.key?(:optional)
  raise ArgumentError, "complain if we have conflicting options"
# Deprecate :required
elsif reflection.options.key?(:required)
  reflection.options[:optional] = !reflection.options[:required]
  ActiveSupport::Deprecation.warn "switch to `optional: #{reflection.options[:optional]}`"
# Set default :optional option
elsif !reflection.options.key?(:optional)
  reflection.options[:optional] = !model.require_belongs_to
end

super

if !reflection.options[:optional]
  model.validates_presence_of …
end
```

File: activerecord/lib/active_record/associations/builder/has_one.rb [hunk 4]
```diff
@@ -17,5 +17,12 @@ def self.valid_dependent_options
     def self.add_destroy_callbacks(model, reflection)
       super unless reflection.options[:through]
     end
+
+    def self.define_validations(model, reflection)
+      super
+      if reflection.options[:required]
+        model.validates_presence_of reflection.name, message: :required
+      end
+    end
```
Comment (jeremy) [inline activerecord/lib/active_record/associations/builder/has_one.rb]: I think it'll be odd to switch to `optional:` on belongs_to but keep `required:` for has_one.

File: activerecord/lib/active_record/associations/builder/belongs_to.rb (hunk 3 above)
Comment (simi) [inline activerecord/lib/active_record/associations/builder/belongs_to.rb]: @jeremy I try to rethink this flow, but deprecation was marked as unnecessary by @dhh here.

File: activerecord/lib/active_record/core.rb [hunk 5]
```diff
@@ -87,6 +87,8 @@ def self.configurations
 
       mattr_accessor :maintain_test_schema, instance_accessor: false
 
+      mattr_accessor :require_belongs_to, instance_accessor: false
+
```
Comment (jeremy) [inline activerecord/lib/active_record/core.rb]: Out of context, I wouldn't know what this "require belongs to" means. Is there a clearer config name?

Comment (simi) [inline activerecord/lib/active_record/core.rb]: yup, it is on the way

File: activerecord/test/cases/associations/required_test.rb [hunk 6]
```diff
@@ -34,7 +34,7 @@ class Child < ActiveRecord::Base
 
   test "required belongs_to associations have presence validated" do
     model = subclass_of(Child) do
-      belongs_to :parent, required: true, inverse_of: false,
+      belongs_to :parent, optional: false, inverse_of: false,
```
Comment (jeremy) [inline activerecord/test/cases/associations/required_test.rb]: If we still support `required: true`, we should still test it.

File: activerecord/lib/active_record/railtie.rb [hunk 7]
```diff
@@ -32,6 +32,7 @@ class Railtie < Rails::Railtie # :nodoc:
 
     config.active_record.use_schema_cache_dump = true
     config.active_record.maintain_test_schema = true
+    config.active_record.require_belongs_to = true
```
Comment (jeremy) [inline activerecord/lib/active_record/railtie.rb]: Since this is in the Railtie, it'll surprisingly affect apps that upgrade to newer Rails. We'll need to generate new apps with it set, but use `false` as the default to preserve existing app behavior.

Comment (simi) [inline activerecord/lib/active_record/railtie.rb]: sure, it is on the way also

**Comment by jeremy** [comment]
Nice work @simi. Pardon the repeated remarks, looks like I missed concurrent feedback too. Thanks!

**Comment by simi** [comment]
Nevermind @jeremy. Thanks for review.

I have just one more question. I made change where `config.active_record.require_belongs_to = true` is generated to `application.rb` for new apps. But it is only one uncommented (unless :skip_active_record) configuration in that file. Should I left it there or should it be better to move into custom initializer?

**Comment by simi** [comment]
I bet it will be better as initializer after inspecting current generated initializers.

**Comment by simi** [comment]
@jeremy @dhh I pushed new commit (I can squash later, just to highlight new code). `config.active_record.belongs_to_required_by_default = true` is now part of `config/initializers/active_record_belongs_to_required_by_default.rb`.

This initializer is generated only when `ActiveRecord` is not skipped and is not added for `rake rails:update`. I think this is compatible enough to skip deprecation.

File: activerecord/lib/active_record/associations/builder/has_one.rb (hunk 4 above)
Comment (dhh) [inline activerecord/lib/active_record/associations/builder/has_one.rb]: I don't think so. belongs_to to me strongly implies a owner/ownee relationship. Doesn't make sense in most cases to say I'm an ownee without a owner. But the other case does make sense. That the owner may or may not have an ownee. (Excuse the made-up word).

**Comment by dhh** [comment]
I think we want the deprecation still in place because we want people to switch over. That is, set config.active_record.belongs_to_required_by_default themselves, and then update with optional or nothing as needed. Because we'll remove config.active_record.belongs_to_required_by_default in Rails 5.x or 6.

**Comment by simi** [comment]
Ahh, ok. Probably I misunderstood you in comments. I'll add deprecation back.

**Comment by simi** [comment]
I mean here - https://github.com/rails/rails/pull/18937#discussion-diff-24725280.

**Comment by dhh** [comment]
Hmm, actually. Good point. Not sure there is a good way to inform about this via a deprecation. I think we’ll have to simply make the deprecation a matter of documentation on the belongs_to method.

**Comment by simi** [comment]
I'll add this into documentation anyway. 

Deprecation can be returned also to https://github.com/rails/rails/pull/18937/files#diff-a26dc1fd56f9dac74db1300aa767378dR24.

**Comment by dhh** [comment]
I think all we need is the documentation updates now.

**Comment by simi** [comment]
I added some really basic documentation note.

screenshot-localhost 3333 2015-02-19 03-17-43

I did a quick read thru guides and I changed validated documentation to optional for belongs_to there.

**Comment by dhh** [comment]
Looks good. We can add further refinements later!

**Comment by simi** [comment]
@jeremy what about https://github.com/rails/rails/pull/18937/files#diff-d853d726a0a949328b3a7fb8f9592c7eR114? Are you still considering it hard to follow?

File: guides/source/association_basics.md [hunk 8]
```diff
@@ -832,7 +832,7 @@ The `belongs_to` association supports these options:
 * `:inverse_of`
 * `:polymorphic`
 * `:touch`
-* `:validate`
+* `:optional`
```
Comment (rafaelfranca) [inline guides/source/association_basics.md]: We still have the `:validate` option. It is not the same as `:required` so we should keep its documentation.

**Comment by rafaelfranca** [comment]
We need to document `belongs_to_required_by_default` in the configuration guide.

Also had you tested the initializer in a new application? I'm not sure if it will work as it is right now because of load order of Active Record.

**Comment by simi** [comment]
@rafaelfranca I reverted my `validate` doc changes and added a note into configuration guide.

**Comment by simi** [comment]
And I tested this via :

```
bin/rails new artest --dev
cd artest
bin/rails g model post
bin/rails g model comment post:belongs_to
bin/rake db:migrate
bin/rails runner "p Comment.new.valid?" #=> false
RAILS_ENV=production bin/rake db:migrate
RAILS_ENV=production bin/rails runner "p Comment.new.valid?" #=> false

```

**Comment by mockdeep** [comment]
Just came across this pull request. I think validating presence of `belongs_to` relationships is going to result in a lot of unnecessary additional queries in codebases, especially if you're validating a number of records before bulk importing or updating them in the database. We've been moving in the direction of instead validating presence of the associated `_id` and letting the foreign key constraint in the database ensure the associated record exists.

**Comment by Silex** [comment]
I feel this change was a bit of a mistake at the design level... validations and relations are different things, and now we get validations done when defining relations.

I understand the "it's convenient" argument tho.

File: guides/source/association_basics.md [hunk 9]
```diff
@@ -956,6 +957,10 @@ end
 
 If you set the `:validate` option to `true`, then associated objects will be validated whenever you save this object. By default, this is `false`: associated objects will not be validated when this object is saved.
 
+##### `:optional`
+
+If you set the `:optional` option to `true`, then associated object will be validated for presence. By default, this is `false`: associated objects will be validated for presence.
```
Comment (Silex) [inline guides/source/association_basics.md]: Typo: "will NOT be validated for presence"

Comment (simi) [inline guides/source/association_basics.md]: **optional:** By default, this is **false**: associated objects will **be** validated for presence.

Comment (Silex) [inline guides/source/association_basics.md]: Look more closely, the sentence says the same thing twice:

```
If you set the `:optional` option to `true`, then associated object 
will be validated for presence. 
By default, this is `false`: associated objects 
will be validated for presence.
```

I took the original sentence and just modified the alignement so it's visually clear they are the same.

Comment (Silex) [inline guides/source/association_basics.md]: Ok, this was fixed in https://github.com/rails/rails/commit/803ef74f9bbcbc71af12060157a996f7b65e24db

**Comment by lekansogunle** [comment]
We are only testing for validation of the presence of the association ID, what if the given ID is invalid. And app behaviour is not to validate the existence of such association?
